File: backend/app/schemas/bankruptcy.py
from __future__ import annotations

from typing import Literal

from pydantic import Field

from app.schemas.common import ApiModel

UserRole = Literal["client", "attorney"]
CaseStatus = Literal[
    "draft",
    "collecting_information",
    "submitted",
    "attorney_review",
    "consultation_scheduled",
    "decision_pending",
    "filing_preparation",
    "closed",
]
Frequency = Literal["weekly", "biweekly", "semimonthly", "monthly", "quarterly", "annual"]
DebtType = Literal["secured", "priority", "unsecured"]
EvidenceStatus = Literal["missing", "requested", "received", "reviewed"]


class HouseholdDto(ApiModel):
    marital_status: str | None = None
    household_size: int = Field(default=1, ge=1, le=20)
    dependents: int = Field(default=0, ge=0, le=20)
    filing_jointly: bool = False
    housing_status: str | None = None
    municipality: str | None = None
    urgent_collection_action: bool = False
    recent_property_transfer: bool = False


class IncomeEntryDto(ApiModel):
    id: str
    category: str
    source: str
    gross_amount: float = Field(ge=0)
    net_amount: float | None = Field(default=None, ge=0)
    frequency: Frequency = "monthly"
    evidence_ids: list[str] = Field(default_factory=list)


class ExpenseEntryDto(ApiModel):
    id: str
    category: str
    description: str
    monthly_amount: float = Field(ge=0)
    essential: bool = True
    evidence_ids: list[str] = Field(default_factory=list)


class DebtEntryDto(ApiModel):
    id: str
    creditor: str
    debt_type: DebtType
    description: str
    balance: float = Field(ge=0)
    monthly_payment: float = Field(default=0, ge=0)
    delinquent_amount: float = Field(default=0, ge=0)
    collateral: str | None = None
    collection_lawsuit: bool = False
    evidence_ids: list[str] = Field(default_factory=list)


class AssetEntryDto(ApiModel):
    id: str
    category: str
    description: str
    estimated_value: float = Field(ge=0)
    loan_balance: float = Field(default=0, ge=0)
    jointly_owned: bool = False
    evidence_ids: list[str] = Field(default_factory=list)


class EvidenceItemDto(ApiModel):
    id: str
    evidence_type: str
    name: str
    status: EvidenceStatus = "received"
    note: str | None = None
    related_entry_ids: list[str] = Field(default_factory=list)


class BankruptcyCaseDto(ApiModel):
    id: str
    owner_user_id: str
    client_name: str
    client_email: str
    client_phone: str | None = None
    preferred_language: str = "es"
    status: CaseStatus = "draft"
    household: HouseholdDto = Field(default_factory=HouseholdDto)
    incomes: list[IncomeEntryDto] = Field(default_factory=list)
    expenses: list[ExpenseEntryDto] = Field(default_factory=list)
    debts: list[DebtEntryDto] = Field(default_factory=list)
    assets: list[AssetEntryDto] = Field(default_factory=list)
    evidence: list[EvidenceItemDto] = Field(default_factory=list)
    client_goal: str | None = None
    attorney_notes: str | None = None


class CaseAnalysisRequestDto(ApiModel):
    case: BankruptcyCaseDto


class CaseAnalysisDto(ApiModel):
    monthly_gross_income: float
    monthly_net_income: float
    monthly_expenses: float
    monthly_cash_flow: float
    total_debt: float
    secured_debt: float
    priority_debt: float
    unsecured_debt: float
    total_asset_value: float
    net_asset_value: float
    completion_score: int
    evidence_score: int
    missing_items: list[str]
    warnings: list[str]
    discussion_points: list[str]
    chapter_7_questions: list[str]
    chapter_13_questions: list[str]
    required_evidence: list[str]
    next_steps: list[str]


AssistantScope = Literal["case", "portfolio"]
"""Which surface the question was asked from — a hint, never an authorization.

The UI already knows whether the user is in the attorney dashboard or inside a
case workspace, and that is cheaper and far more reliable than classifying every
sentence with keyword heuristics. So the client says where it is.

What it cannot do is grant itself anything. The server pairs this with the
authenticated role: `portfolio` from a client session is ignored, and the
collection itself is always resolved server-side from the session identity. The
hint chooses *which* authorized scope to build, never *what* is in it.

Defaults to `case`, so an older client that sends nothing keeps its current
behaviour and never accidentally opens a wider surface.
"""


class GuidanceRequestDto(ApiModel):
    case: BankruptcyCaseDto
    message: str
    role: UserRole
    locale: str = "es"
    assistant_scope: AssistantScope = "case"

File: frontend/src/types/bankruptcy.ts
import type { UserRole } from "./api";

export type CaseStatus =
  | "draft"
  | "collecting_information"
  | "submitted"
  | "attorney_review"
  | "consultation_scheduled"
  | "decision_pending"
  | "filing_preparation"
  | "closed";

export type Frequency =
  | "weekly"
  | "biweekly"
  | "semimonthly"
  | "monthly"
  | "quarterly"
  | "annual";
export type DebtType = "secured" | "priority" | "unsecured";
export type EvidenceStatus = "missing" | "requested" | "received" | "reviewed";
export type EntryKind = "income" | "expense" | "debt" | "asset" | "evidence";

export interface Household {
  maritalStatus?: string;
  householdSize: number;
  dependents: number;
  filingJointly: boolean;
  housingStatus?: string;
  municipality?: string;
  urgentCollectionAction: boolean;
  recentPropertyTransfer: boolean;
}

export interface IncomeEntry {
  id: string;
  category: string;
  source: string;
  grossAmount: number;
  netAmount?: number;
  frequency: Frequency;
  evidenceIds: string[];
}

export interface ExpenseEntry {
  id: string;
  category: string;
  description: string;
  monthlyAmount: number;
  essential: boolean;
  evidenceIds: string[];
}

export interface DebtEntry {
  id: string;
  creditor: string;
  debtType: DebtType;
  description: string;
  balance: number;
  monthlyPayment: number;
  delinquentAmount: number;
  collateral?: string;
  collectionLawsuit: boolean;
  evidenceIds: string[];
}

export interface AssetEntry {
  id: string;
  category: string;
  description: string;
  estimatedValue: number;
  loanBalance: number;
  jointlyOwned: boolean;
  evidenceIds: string[];
}

export interface EvidenceItem {
  id: string;
  evidenceType: string;
  name: string;
  status: EvidenceStatus;
  note?: string;
  relatedEntryIds: string[];
}

export interface ChatMessage {
  id: string;
  role: "assistant" | "user";
  content: string;
  createdAt: string;
}

export interface TimelineEvent {
  id: string;
  stage: string;
  title: string;
  description: string;
  status: "complete" | "current" | "upcoming";
  createdAt: string;
}

export interface BankruptcyCase {
  id: string;
  ownerUserId: string;
  clientName: string;
  clientEmail: string;
  clientPhone?: string;
  preferredLanguage: string;
  status: CaseStatus;
  household: Household;
  incomes: IncomeEntry[];
  expenses: ExpenseEntry[];
  debts: DebtEntry[];
  assets: AssetEntry[];
  evidence: EvidenceItem[];
  clientGoal?: string;
  attorneyNotes?: string;
  assignedAttorneyName?: string;
  createdAt: string;
  updatedAt: string;
  submittedAt?: string;
  messages: ChatMessage[];
  timeline: TimelineEvent[];
}

export interface CaseAnalysis {
  monthly_gross_income: number;
  monthly_net_income: number;
  monthly_expenses: number;
  monthly_cash_flow: number;
  total_debt: number;
  secured_debt: number;
  priority_debt: number;
  unsecured_debt: number;
  total_asset_value: number;
  net_asset_value: number;
  completion_score: number;
  evidence_score: number;
  missing_items: string[];
  warnings: string[];
  discussion_points: string[];
  chapter_7_questions: string[];
  chapter_13_questions: string[];
  required_evidence: string[];
  next_steps: string[];
}

/**
 * Assistant contract, 4.0.0 (ADR 0002).
 *
 * Mirrors `backend/app/ai/contracts/assistant_response.py`. Read-only action
 * vocabulary — write actions are phase 2 and require the signed
 * server-side confirmation flow, which is why `requires_confirmation` is
 * already carried but nothing sets it yet.
 */
export type AssistantActionType =
  | "open_page"
  | "show_details"
  | "upload_document"
  | "request_document"
  | "ask";

export interface AssistantAction {
  id: string;
  action_type: AssistantActionType;
  /** A case-workspace section. Validated against ASSISTANT_ACTION_RESOURCES
   * before anything is rendered or navigated to — see assistantActions.ts. */
  resource: string;
  resource_id?: string | null;
  label: string;
  icon: string;
  payload: Record<string, unknown>;
  requires_confirmation: boolean;
}

export interface AssistantCard {
  card_type: string;
  title: string;
  description?: string | null;
  data: Record<string, unknown>;
}

export interface AssistantResponse {
  language: string;
  message: string;
  /** Which specialist produced the answer, or "deterministic" when the agent
   * layer could not answer. */
  handled_by: string;
  actions: AssistantAction[];
  cards: AssistantCard[];
  warnings: string[];
  requires_attorney_review: boolean;
  /** True when the answer came from the deterministic fallback rather than
   * the agent layer. Surfaced so the UI can say so instead of presenting a
   * rule-based draft as a model response. */
  degraded: boolean;
  disclaimer: string;
}

export type EntrySubmission =
  | { kind: "income"; value: IncomeEntry }
  | { kind: "expense"; value: ExpenseEntry }
  | { kind: "debt"; value: DebtEntry }
  | { kind: "asset"; value: AssetEntry }
  | { kind: "evidence"; value: EvidenceItem };

export interface ApiBankruptcyCase {
  id: string;
  owner_user_id: string;
  client_name: string;
  client_email: string;
  client_phone?: string;
  preferred_language: string;
  status: CaseStatus;
  household: {
    marital_status?: string;
    household_size: number;
    dependents: number;
    filing_jointly: boolean;
    housing_status?: string;
    municipality?: string;
    urgent_collection_action: boolean;
    recent_property_transfer: boolean;
  };
  incomes: Array<{
    id: string;
    category: string;
    source: string;
    gross_amount: number;
    net_amount?: number;
    frequency: Frequency;
    evidence_ids: string[];
  }>;
  expenses: Array<{
    id: string;
    category: string;
    description: string;
    monthly_amount: number;
    essential: boolean;
    evidence_ids: string[];
  }>;
  debts: Array<{
    id: string;
    creditor: string;
    debt_type: DebtType;
    description: string;
    balance: number;
    monthly_payment: number;
    delinquent_amount: number;
    collateral?: string;
    collection_lawsuit: boolean;
    evidence_ids: string[];
  }>;
  assets: Array<{
    id: string;
    category: string;
    description: string;
    estimated_value: number;
    loan_balance: number;
    jointly_owned: boolean;
    evidence_ids: string[];
  }>;
  evidence: Array<{
    id: string;
    evidence_type: string;
    name: string;
    status: EvidenceStatus;
    note?: string;
    related_entry_ids: string[];
  }>;
  client_goal?: string;
  attorney_notes?: string;
}

export interface WorkspaceState {
  cases: BankruptcyCase[];
}

export interface GuidanceRequest {
  case: ApiBankruptcyCase;
  message: string;
  role: UserRole;
  locale: string;
}
